File: chap12/fig12-8.R
library(dplyr)
library(ggplot2)
library(ggrepel)

d <- read.csv('input/data-map-temperature.csv')
d_name <- read.csv('input/data-map-JIS.csv', header=FALSE, fileEncoding='UTF-8')
colnames(d_name) <- c('prefID', 'name')

fit <- readRDS('output/result-model12-5.RDS')
f_ms <- fit$draws('f', format='matrix')
qua <- apply(f_ms, 2, quantile, probs=c(0.025, 0.25, 0.50, 0.75, 0.975))
d_est <- data.frame(d_name, Y=d$Y, t(qua), check.names=FALSE) %>% 
  mutate(residual=Y-`50%`)
top_residual <- which(rank(-abs(d_est$residual))==1)

p <- ggplot(data=d_est, aes(x=Y, y=`50%`, ymin=`2.5%`, ymax=`97.5%`)) +
  theme_bw(base_size=18) +
  theme(legend.key.height=grid::unit(2.5,'line')) +
  coord_fixed(ratio=1, xlim=c(9, 24), ylim=c(9, 24)) +
  geom_label_repel(
    data=d_est[top_residual, ], aes(label=name),
    nudge_x = -0.5,
    nudge_y = 0.15,
    box.padding=grid::unit(2, 'lines')
  ) +
  geom_pointrange(size=0.8, shape=21, fill='white') +
  geom_abline(aes(slope=1, intercept=0), color='black', alpha=3/5, linetype='31') +
  labs(x='Observed', y='Predicted') +
  scale_x_continuous(breaks=seq(0, 25, 5)) +
  scale_y_continuous(breaks=seq(0, 25, 5))
ggsave(p, file='output/fig12-8-left.png', dpi=300, w=4, h=4)


noise_ms <- t(replicate(nrow(f_ms), d$Y)) - f_ms
# noise_ms <- einsum::einsum('n,m->mn', d$Y, rep(1, nrow(f_ms))) - f_ms

d_mode <- apply(noise_ms, 2, function(x) {
  dens <- density(x)
  mode_i <- which.max(dens$y)
  mode_x <- dens$x[mode_i]
  mode_y <- dens$y[mode_i]
  c(mode_x, mode_y)
}) %>% 
  t() %>% 
  data.frame() %>% 
  magrittr::set_colnames(c('X', 'Y'))

s_dens <- density(fit$draws('s_y', format='matrix'))
s_MAP <- s_dens$x[which.max(s_dens$y)]
bw <- 0.05
p <- ggplot(data=d_mode, aes(x=X)) +
  theme_bw(base_size=18) +
  geom_histogram(aes(y=..density..), binwidth=bw, color='black', fill='white') +
  geom_density(alpha=0.4, color=NA, fill='gray20') +
  geom_rug(sides='b') +
  stat_function(fun=function(x) dnorm(x, mean=0, sd=s_MAP), linetype='dashed') +
  labs(x='value', y='density') +
  xlim(range(density(d_mode$X)$x))
ggsave(file='output/fig12-8-right.png', plot=p, dpi=300, w=4, h=4)

File: chap12/fig12-8.py
import pandas
import cmdstanpy
import numpy as np
from scipy import stats
from plotnine import *

d = pandas.read_csv('input/data-map-temperature.csv')
N = len(d)
d_name = pandas.read_csv('input/data-map-JIS.csv', header=None)
d_name.columns = ['prefID', 'name']
fit = cmdstanpy.from_csv('output/result-model12-5')
f_ms = fit.stan_variable(var='f')

qua = np.quantile(f_ms, [0.025, 0.25, 0.50, 0.75, 0.975], axis=0)
d_qua = pandas.DataFrame(qua.T, columns=['2.5%', '25%', '50%', '75%', '97.5%'])
d_est = pandas.concat([d_name, d.Y, d_qua], axis=1)
d_est['residual'] = d_est.Y - d_est['50%']
top_residual = d_est.iloc[np.argmax(np.abs(d_est.residual))]

p = (ggplot(d_est, aes(x='Y', y='50%', ymin='2.5%', ymax='97.5%'))
    + theme_bw(base_size=18)
    + coord_fixed(ratio=1, xlim=[9, 24], ylim=[9, 24])
    + geom_pointrange(size=0.8, shape='x', fill='white')
    + geom_abline(aes(slope=1, intercept=0), color='black', alpha=3/5, linetype='dashed')
    + labs(x='Observed', y='Predicted')
    + scale_x_continuous(breaks=np.arange(0, 30, 5))
    + scale_y_continuous(breaks=np.arange(0, 30, 5))
)
p.save(filename='output/fig12-8-left.py.png', dpi=300, width=4, height=4)

N_ms = len(f_ms)
noise_ms = np.repeat(d.Y.to_numpy()[None, :], N_ms, axis=0) - f_ms
# noise_ms = np.einsum('n,m->mn', d.Y, np.repeat(1, N_ms)) - f_ms

mode_list = []
for i in range(N):
    density = stats.gaussian_kde(noise_ms[:,i])
    xs = np.linspace(-1.8, 0.5, 201)
    ys = density(xs)
    mode_i = np.argmax(ys)
    mode_x = xs[mode_i]
    mode_y = ys[mode_i]
    mode_list.append([mode_x, mode_y])

d_mode = pandas.DataFrame(mode_list, columns=['X', 'Y'])

s_dens = stats.gaussian_kde(fit.stan_variable(var='s_y'))
xs = np.linspace(-1.8, 0.5, 201)
ys = s_dens(xs)
s_MAP = xs[np.argmax(ys)]
bw = 0.05

p = (ggplot(d_mode, aes(x='X'))
    + theme_bw(base_size=18)
    + geom_histogram(aes(y='..density..'), binwidth=bw, color='black', fill='white')
    + geom_density(fill='gray', alpha=0.5)
    + geom_rug(sides='b')
    + stat_function(fun=stats.norm.pdf, args=dict(loc=0, scale=s_MAP), linetype='dashed')
    + labs(x='value', y='density')
    + xlim(-1.8, 0.5)
) 
p.save(filename='output/fig12-8-right.py.png', dpi=300, width=4, height=4)
